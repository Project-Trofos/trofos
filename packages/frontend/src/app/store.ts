import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
