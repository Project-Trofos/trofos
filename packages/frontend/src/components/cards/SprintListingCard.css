li.sprint-card-container {
  background: #fff;
  border: 1px solid #eee;
  padding: 0;
}

.sprint-card-inner-container {
  display: flex;
}

.sprint-card-container > div {
  width: 100%;
}

.sprint-card-name {
  display: flex;
  flex: 1 1 0%;
}

.sprint-card-setting-icon {
  padding: 0 6px 0 10px;
}
