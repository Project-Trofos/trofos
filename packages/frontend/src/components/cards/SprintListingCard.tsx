import React, { useRef, useState } from 'react';
import { Button, Collapse, message, Tour } from 'antd';
import { BookOutlined } from '@ant-design/icons';
import dayjs from 'dayjs';
import { useNavigate, useParams } from 'react-router-dom';
import { useUpdateSprintMutation } from '../../api/socket/sprintHooks';
import { Sprint } from '../../api/sprint';
import BacklogList from '../lists/BacklogList';
import SprintMenu from '../dropdowns/SprintMenu';
import StrictModeDroppable from '../dnd/StrictModeDroppable';
import SprintNotesModal from '../modals/SprintNotesModal';
import './SprintListingCard.css';
import type { TourProps } from 'antd';

function SprintListingCard(props: SprintListingCardProps): JSX.Element {
  const { sprint, setSprint, setIsModalVisible } = props;
  const { Panel } = Collapse;
  const navigate = useNavigate();

  const [isNotesOpen, setIsNotesOpen] = useState(false);

  const params = useParams();
  const projectId = Number(params.projectId);

  const [updateSprint] = useUpdateSprintMutation();

  const handleSprintOnClick = (e: React.MouseEvent<HTMLDivElement>) => {
    setSprint(sprint);
    setIsModalVisible(true);
  };

  const handleSprintStatusUpdate = async (
    updatedStatus: 'upcoming' | 'current' | 'completed',
    e: React.MouseEvent<HTMLElement, MouseEvent>,
  ) => {
    e.stopPropagation();

    const payload = {
      status: updatedStatus,
      sprintId: sprint.id,
      projectId,
    };

    try {
      await updateSprint(payload).unwrap();
      message.success('Sprint updated');
      console.log('Success');
    } catch (err: any) {
      message.error(err?.data?.error || 'Failed to update sprint');
      console.error(err);
    }
  };

  const renderSprintStatusButton = () => {
    switch (sprint.status) {
      case 'upcoming':
        return <Button onClick={(e) => handleSprintStatusUpdate('current', e)}>Start Sprint</Button>;
      case 'current':
        return <Button onClick={(e) => handleSprintStatusUpdate('completed', e)}>Complete Sprint</Button>;
      case 'completed':
        return <Button onClick={(e) => handleSprintStatusUpdate('current', e)}>Reopen Sprint</Button>;
      default:
        // eslint-disable-next-line react/jsx-no-useless-fragment
        return <></>;
    }
  };

  const navigateToRetrospective = (sprintId: number) => {
    navigate(`./${sprintId}/retrospective`);
  };

  const navigateToBoard = (sprintId: number) => {
    navigate(`../board/${sprintId}`);
  };

  // const openSprintNotesModal = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
  //   e.stopPropagation();
  //   setIsNotesOpen(true);
  // };
  // TODO: revert when remove tour
  const openSprintNotesModal = () => {
    setIsNotesOpen(true);
  }

  // TODO: delete following when remove tour
  const sprintNotesButtonRef = useRef(null);
  const [isTourOpen, setIsTourOpen] = useState<boolean>(false);
  const tourSteps: TourProps['steps'] = [
    {
      title: "This version of sprint notes will be deprecated soon!",
      description: "We have concurrent sprint notes available in the sprint board",
      cover: (
        <Button 
          onClick={() => {
            navigateToBoard(sprint.id);
          }}
        >
          Click to view new notes
        </Button>
      ),
      target: () => sprintNotesButtonRef.current,
      nextButtonProps: {
        children: (
          <>
            View deprecated notes
          </>
        ),
        onClick: () => {
          setIsTourOpen(false);
          openSprintNotesModal();
        }
      },
      placement: 'leftTop',
      style: {
        border: "2px solid",
        borderRadius: "8px",
        boxShadow: "0 4px 10px rgba(0,0,0,0.1)"
      }
    }
  ];
  // TODO: delete above when remove tour

  return (
    <>
      <StrictModeDroppable droppableId={sprint.id.toString()}>
        {(provided) => (
          <div ref={provided.innerRef} {...provided.droppableProps}>
            <Collapse>
              <Panel
                key={sprint.id}
                header={
                  <div className="sprint-card-inner-container">
                    <div className="sprint-card-name">{sprint.name}</div>
                    <div className="sprint-status-button">{renderSprintStatusButton()}</div>
                    {sprint.status !== 'current' && (
                      <div className="sprint-card-button">
                        <Button onClick={() => navigateToBoard(sprint.id)}>View Board</Button>
                      </div>
                    )}
                    {(sprint.status === 'completed' || sprint.status === 'closed') && (
                      <div className="sprint-card-button">
                        <Button onClick={() => navigateToRetrospective(sprint.id)}>Retrospective</Button>
                      </div>
                    )}
                    {sprint.start_date && sprint.end_date && (
                      <div>{`${dayjs(sprint.start_date).format('DD/MM/YYYY')} - ${dayjs(sprint.end_date).format(
                        'DD/MM/YYYY',
                      )}`}</div>
                    )}
                    <div className="sprint-card-notes-icon">
                      {/* <BookOutlined onClick={openSprintNotesModal} style={{ fontSize: '18px' }} /> */}
                      <BookOutlined onClick={() => {setIsTourOpen(true)}} style={{ fontSize: '18px' }} ref={sprintNotesButtonRef}/>
                    </div>
                    <div className="sprint-card-setting-icon" onClick={(e) => e.stopPropagation()}>
                      <SprintMenu
                        sprintId={sprint.id}
                        projectId={projectId}
                        handleSprintOnClick={handleSprintOnClick}
                      />
                    </div>
                  </div>
                }
              >
                <BacklogList backlogs={sprint.backlogs} />
                {provided.placeholder}
              </Panel>
            </Collapse>
          </div>
        )}
      </StrictModeDroppable>
      {isNotesOpen && <SprintNotesModal isOpen={isNotesOpen} setIsOpen={setIsNotesOpen} sprintId={sprint.id} />}
      {/* TODO: delete the following when remove tour */}
      <Tour
        open={isTourOpen}
        onClose={() => {
          setIsTourOpen(false);
        }}
        steps={tourSteps}
        mask={false}
      />
    </>
  );
}

type SprintListingCardProps = {
  sprint: Sprint;
  setSprint(sprint: Sprint): void;
  setIsModalVisible(isVisible: boolean): void;
};

export default SprintListingCard;
