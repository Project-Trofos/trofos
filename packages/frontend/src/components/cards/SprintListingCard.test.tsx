import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { DragDropContext } from 'react-beautiful-dnd';
import SprintListingCard from './SprintListingCard';
import store from '../../app/store';
import type { Sprint } from '../../api/sprint';

describe('SprintCard test', () => {
  const mockSprint: Sprint = {
    id: 111,
    name: 'Sprint 111',
    duration: 1,
    start_date: '2022-10-09 07:03:56',
    end_date: '2022-10-16 07:03:56',
    project_id: 123,
    goals: 'Some test goals',
    status: 'upcoming',
    backlogs: [],
  };

  const mockSprintListingCardProps = {
    setSprint: vi.fn(),
    setIsModalVisible: vi.fn(),
  };

  const { baseElement } = render(
    <BrowserRouter>
      <Provider store={store}>
        <DragDropContext onDragEnd={vi.fn()}>
          <SprintListingCard
            sprint={mockSprint}
            setIsModalVisible={mockSprintListingCardProps.setIsModalVisible}
            setSprint={mockSprintListingCardProps.setSprint}
          />
        </DragDropContext>
      </Provider>
    </BrowserRouter>,
  );

  it('renders sprint card with correct details', () => {
    expect(screen.getByText('Sprint 111')).toBeInTheDocument();
    expect(screen.getByText('09/10/2022 - 16/10/2022')).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
