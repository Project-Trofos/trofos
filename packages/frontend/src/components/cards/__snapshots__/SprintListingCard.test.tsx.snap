// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintCard test renders sprint card with correct details 1`] = `
<body>
  <div>
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="111"
    >
      <div
        class="ant-collapse ant-collapse-icon-position-start"
      >
        <div
          class="ant-collapse-item"
        >
          <div
            aria-disabled="false"
            aria-expanded="false"
            class="ant-collapse-header"
            role="button"
            tabindex="0"
          >
            <div
              class="ant-collapse-expand-icon"
            >
              <span
                aria-label="right"
                class="anticon anticon-right ant-collapse-arrow"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </div>
            <span
              class="ant-collapse-header-text"
            >
              <div
                class="sprint-card-inner-container"
              >
                <div
                  class="sprint-card-name"
                >
                  Sprint 111
                </div>
                <div
                  class="sprint-status-button"
                >
                  <button
                    class="ant-btn ant-btn-default"
                    type="button"
                  >
                    <span>
                      Start Sprint
                    </span>
                  </button>
                </div>
                <div>
                  09/10/2022 - 16/10/2022
                </div>
                <div
                  class="sprint-card-notes-icon"
                >
                  <span
                    aria-label="book"
                    class="anticon anticon-book"
                    role="img"
                    style="font-size: 18px;"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="book"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-260 72h96v209.9L621.5 312 572 347.4V136zm220 752H232V136h280v296.9c0 3.3 1 6.6 3 9.3a15.9 15.9 0 0022.3 3.7l83.8-59.9 81.4 59.4c2.7 2 6 3.1 9.4 3.1 8.8 0 16-7.2 16-16V136h64v752z"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="sprint-card-setting-icon"
                >
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center ant-dropdown-trigger sprint-menu-dropdown"
                  >
                    <div
                      class="ant-space-item"
                    >
                      <span
                        aria-label="setting"
                        class="anticon anticon-setting"
                        role="img"
                        style="font-size: 18px;"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="setting"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M924.8 625.7l-65.5-56c3.1-19 4.7-38.4 4.7-57.8s-1.6-38.8-4.7-57.8l65.5-56a32.03 32.03 0 009.3-35.2l-.9-2.6a443.74 443.74 0 00-79.7-137.9l-1.8-2.1a32.12 32.12 0 00-35.1-9.5l-81.3 28.9c-30-24.6-63.5-44-99.7-57.6l-15.7-85a32.05 32.05 0 00-25.8-25.7l-2.7-.5c-52.1-9.4-106.9-9.4-159 0l-2.7.5a32.05 32.05 0 00-25.8 25.7l-15.8 85.4a351.86 351.86 0 00-99 57.4l-81.9-29.1a32 32 0 00-35.1 9.5l-1.8 2.1a446.02 446.02 0 00-79.7 137.9l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.3 56.6c-3.1 18.8-4.6 38-4.6 57.1 0 19.2 1.5 38.4 4.6 57.1L99 625.5a32.03 32.03 0 00-9.3 35.2l.9 2.6c18.1 50.4 44.9 96.9 79.7 137.9l1.8 2.1a32.12 32.12 0 0035.1 9.5l81.9-29.1c29.8 24.5 63.1 43.9 99 57.4l15.8 85.4a32.05 32.05 0 0025.8 25.7l2.7.5a449.4 449.4 0 00159 0l2.7-.5a32.05 32.05 0 0025.8-25.7l15.7-85a350 350 0 0099.7-57.6l81.3 28.9a32 32 0 0035.1-9.5l1.8-2.1c34.8-41.1 61.6-87.5 79.7-137.9l.9-2.6c4.5-12.3.8-26.3-9.3-35zM788.3 465.9c2.5 15.1 3.8 30.6 3.8 46.1s-1.3 31-3.8 46.1l-6.6 40.1 74.7 63.9a370.03 370.03 0 01-42.6 73.6L721 702.8l-31.4 25.8c-23.9 19.6-50.5 35-79.3 45.8l-38.1 14.3-17.9 97a377.5 377.5 0 01-85 0l-17.9-97.2-37.8-14.5c-28.5-10.8-55-26.2-78.7-45.7l-31.4-25.9-93.4 33.2c-17-22.9-31.2-47.6-42.6-73.6l75.5-64.5-6.5-40c-2.4-14.9-3.7-30.3-3.7-45.5 0-15.3 1.2-30.6 3.7-45.5l6.5-40-75.5-64.5c11.3-26.1 25.6-50.7 42.6-73.6l93.4 33.2 31.4-25.9c23.7-19.5 50.2-34.9 78.7-45.7l37.9-14.3 17.9-97.2c28.1-3.2 56.8-3.2 85 0l17.9 97 38.1 14.3c28.7 10.8 55.4 26.2 79.3 45.8l31.4 25.8 92.8-32.9c17 22.9 31.2 47.6 42.6 73.6L781.8 426l6.5 39.9zM512 326c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm79.2 255.2A111.6 111.6 0 01512 614c-29.9 0-58-11.7-79.2-32.8A111.6 111.6 0 01400 502c0-29.9 11.7-58 32.8-79.2C454 401.6 482.1 390 512 390c29.9 0 58 11.6 79.2 32.8A111.6 111.6 0 01624 502c0 29.9-11.7 58-32.8 79.2z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="rbd-announcement-0"
    style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
  />
  <div
    id="rbd-hidden-text-0-hidden-text-0"
    style="display: none;"
  >
    
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

  </div>
</body>
`;
